tests/test_misc.lua

```diff
@@ -541,4 +542,382 @@ T['zoom()']['respects `config` argument'] = function()
   child.expect_screenshot()
 end
 
+T['setup_restore_cursor()'] = new_set()
+
+local setup_restore_cursor = function(...) child.lua('MiniMisc.setup_restore_cursor(...)', { ... }) end
+
+local restore_cursor_default_opts = {
+  ignore_buftype = { 'nofile' },
+  ignore_filetype = { 'gitcommit' },
+  center = true,
+}
+
+T['setup_restore_cursor()']['autocmd'] = function()
+  setup_restore_cursor()
+
+  eq(child.lua_get([[#vim.api.nvim_get_autocmds({ group = 'MiniMiscRestoreCursor' })]]) > 0, true)
+end
+
+T['setup_restore_cursor()']['opts'] = function()
+  setup_restore_cursor(restore_cursor_default_opts)
+
+  eq(child.lua_get([[#vim.api.nvim_get_autocmds({ group = 'MiniMiscRestoreCursor' })]]) > 0, true)
+end
+
+T['setup_restore_cursor()']['invalid opts'] = function()
+  expect.error(function() setup_restore_cursor({ ignore_buftype = true }) end, '`opts.ignore_buftype`.*array')
+  expect.error(function() setup_restore_cursor({ ignore_filetype = true }) end, '`opts.ignore_filetype`.*array')
+  expect.error(function() setup_restore_cursor({ center = '' }) end, '`opts.center`.*boolean')
+end
+
+local cursor_dir_relpath = 'tests/dir-misc/cursor'
+local cursor_dir_path = vim.fn.fnamemodify(cursor_dir_relpath, ':p')
+local cursor_shada_file = string.format("%s/cursor.shada", cusor_dir_path)
+
+local cursor_cleanup = function()
+  child.fn.delete(cursor_dir_path, 'rf')
+end
+
+T['restore_cursor()'] = new_set({
+  hooks = {
+    pre_case = function()
+      child.setup()
+
+      -- Ensure `require('mini.sessions')` is always possible (which might not be
+      -- the case with all present `cd` calls)
+      child.cmd('set rtp+=' .. project_root)
+
+      -- Ensure always identical starting current directory
+      cd(project_root)
+
+      -- Ensure directory structure invariants
+      cursor_cleanup()
+
+      -- Make all showed messages full width
+      child.o.cmdheight = 10
+
+      -- Load module
+      load_module()
+
+      -- Ensure clear message history
+      child.cmd('messages clear')
+    end,
+    post_once = function()
+      cursor_cleanup()
+      child.stop()
+    end,
+  },
+})
+
+local restore_cursor = function(...) child.lua('MiniMisc.restore_cursor(...)', { ... }) end
+local test_file_cursor = make_abspath('tests/dir-misc/cursor_restore.lua')
+
+T['restore_cursor()']['opts'] = function()
+  local opts = restore_cursor_default_opts
+
+  restore_cursor(opts)
+end
+
+local cursor_jump_line = function(line)
+  child.cmd(string.format('%u', line))
+  eq(child.fn.line('.'), line)
+end
+
+local write_shada_file = function()
+  child.fn.mkdir(cursor_dir_path)
+  child.o.shadafile = ''
+  child.cmd('wshada! ' .. cursor_shada_file)
+  -- verify that shada file has been written
+  local stat = vim.loop.fs_stat(cursor_shada_file)
+  eq(stat.type, 'file')
+  eq(stat.size > 1, true)
```

Comment:
This is more helpful during debugging, not actual tests. Should be safe to remove.

Reply:
Done.